raft/src/main/java/org/apache/kafka/raft/FileBasedStateStore.java

```diff
@@ -0,0 +1,178 @@
+
+    public FileBasedStateStore(final File stateFile) {
+        this.stateFile = stateFile;
+    }
+
+    @Override
+    public void initialize(OffsetAndEpoch logEndOffsetAndEpoch) throws IOException {
+        cachedState = loadState(stateFile, logEndOffsetAndEpoch);
+        log.info("Loaded quorum state from local: {}", cachedState);
+    }
+
+    private QuorumStateData loadState(File file, final OffsetAndEpoch logEndOffsetAndEpoch) throws IOException {
+        try (final BufferedReader reader = Files.newBufferedReader(file.toPath())) {
+            final ObjectMapper objectMapper = new ObjectMapper();
+            JsonNode readNode = objectMapper.readTree(readLine(reader));
+
+            if (!(readNode instanceof ObjectNode)) {
+                throw new IllegalStateException("Deserialized node is not an object node");
+            }
+            final ObjectNode dataObject = (ObjectNode) readNode;
+            final short dataVersion = dataObject.get(DATA_VERSION).shortValue();
+
+            QuorumStateData data = new QuorumStateData();
+            data.fromJson(dataObject, dataVersion);
+            return data;
+        } catch (final NoSuchFileException e) {
+            return recoverStateFromLog(file, logEndOffsetAndEpoch);
+        } catch (final Exception e) {
+            // For other exception during state file loading, we could assume the file is corrupted already
+            // and should be cleaned up.
+            delete();
+            log.info("Local quorum file corrupted, deleting {}", file.toPath());
+            return recoverStateFromLog(file, logEndOffsetAndEpoch);
+        }
+    }
+
+    private QuorumStateData recoverStateFromLog(File file, final OffsetAndEpoch logEndOffsetAndEpoch) throws IOException {
+        QuorumStateData stateFromMetadataLog = new QuorumStateData();
+        incorporateElectionState(stateFromMetadataLog,
+            ElectionState.withUnknownLeader(logEndOffsetAndEpoch.epoch));
+        writeElectionStateToFile(file, stateFromMetadataLog);
+        return stateFromMetadataLog;
+    }
+
+    /**
+     * Reads the election state from cached quorum data.
+     */
+    @Override
+    public ElectionState readElectionState() {
+        return new ElectionState(
+            cachedState.leaderEpoch(),
+            cachedState.leaderId() == UNKNOWN_LEADER_ID ? OptionalInt.empty() :
+            OptionalInt.of(cachedState.leaderId()),
+            cachedState.votedId() == NOT_VOTED ? OptionalInt.empty() :
+            OptionalInt.of(cachedState.votedId()));
+    }
+
+    @Override
+    public void writeElectionState(ElectionState latest) throws IOException {
+        incorporateElectionState(cachedState, latest);
+        writeElectionStateToFile(stateFile, cachedState);
+    }
+
+    private void incorporateElectionState(QuorumStateData state, ElectionState latest) {
+        state.setLeaderEpoch(latest.epoch)
+            .setVotedId(latest.hasVoted() ? latest.votedId() : NOT_VOTED)
```

Comment:
It's still a bit concerning to me that, upon losing the state quorum file we could potentially vote for different leaders within a single epoch. An idea to cope with this scenario on top of my head:

* Upon starting up, the follower does not try to become a voter (or at least mark it self as not being able to vote for anyone) if it cannot find the quorum state file. Similarly if a voter downgrade to an observer, it deletes its local quorum-state file. So that only current-voters would have quorum file, and as long as it exists it is always up-to-date.

* Ideally a voter would not vote for anyone until she learns the latest epoch, but in theory no one can tell for sure if it has learned the latest epoch because even if your fetch-quorum-records returns OK, it might be from a zombie leader as well. So taking a step-back, we just forbid a voter started without the state-quorum to vote for anyone after, say, 2 * (election.timeout + jitter.ms), then even if it has voted for someone on a specific epoch, that vote has either made someone to become the leader, or that vote has been void since the candidate already knew it did not win the majority for that epoch. In either case, the voted epoch she has participated but forgot about, should have ended by then.

* For bootstrapping where we do not have state-quorum files anyways and starting epoch = 0, we are unnecessarily delaying the form of the quorum by 2 * election.timeout. I think this is not a very big deal in order to be on the safer side.

WDYT?

Reply:
I am not sure. The edge case we're trying to protect is extremely unlikely and even this suggestion is not bulletproof. So would we reduce the odds of this happening from 1 in a billion to 1 in 10 billion? Would that be enough to make it worthwhile? On the other hand, _everyone_ will feel the pain of slow bootstrapping.. 

If we really want to solve this problem, we should be looking at byzantine failure models which allow "lying" nodes like this. Practically speaking, I think the best we can do is detect inconsistent states and fail fast. For example, when the leader asks a follower to truncate below its last known high watermark, we can fail. As long as you have a check like that in place, then the potential damage from a split brain is reduced.

I am not sure the answer though. I know that Zookeeper has definitely encountered problems where nodes recovering from a bad disk state have caused split brains and data loss. So we should definitely be thinking about this.

Reply:
Fair enough, let's pound this idea and keep in our pocket of now --- if we did observe it is already very rare and good tooling would be sufficient, then we should not be over-protective and sacrifice bootstrapping latency.

Reply:
If we follow @hachikuji suggestion and port the current fetch protection mechanisms here, what's the recovery path? How does Kafka replication survive such failures when they hit fatal truncation under HW exception?

Reply:
We always have the last resort to truncate all the logs and fetch / bootstrap from the very beginning.

Reply:
Another thought I have is that whether this scenario becomes severe only in the pull based model. In a typical 3 node A, B, C quorum, where A is the leader and B, C are followers. Only B is up to date of HW. Then node A crashes, and the local disk is corrupted as well. When starting up, A has no state so it will participate in the vote of C and elect C as the new leader. However, in this case, we could assume A is already a failed copy, so we are left with B, C to elect a new leader at this time. First of all, B will never vote for C, as it has a longer log. But here is the tricky thing, we don't compare offset in the `BeginQuorumEpoch` Protocol, which puts us in the risk of truncating. What if we enforce end offset comparison for `BeginQuorumEpoch` protocol as well, so that no way a less up-to-date member could be elected as leader, by rejecting the request from B and let C becomes a follower.

So in B's perspective, what if we have a way to reject a `BeginQuorumEpoch` request when:
1. the last epoch was the same as B's current epoch
2. this member has been rejected before because its log is not sufficiently catching up

After the rejection takes place, we could either fail fast as we don't have a way to win election any more until a human intervention by wiping out C's quorum-state file as well. To auto recover in this case, I could only think of incrementing B's epoch by 2 to catch up with C's epoch, but this could also bring in unknown risk.

I know this is not a discussed case in the literature, so just put out this random thought for discussion, and see if you feel this could potentially fix the problem.

Reply:
After a bit of thought, I think this is a similar approach to the high watermark truncation protection. Maybe we should not over-complicate the logic and just use what we already have for Kafka replication.